### ISSUE TITLE
TOB-K8S-026: Directory traversal of host logs running kube-apiserver and kubelet

### ISSUE THREAD

**Issue by cji** [kubernetes/kubernetes#81125]
This issue was reported in the [Kubernetes Security Audit Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Description**
The kube-apiserver runs within the Control Plane of a Kubernetes cluster, acting as the central authority on cluster state. Part of its functionality is serving files from /var/log on the /logs[/{logpath:*}] routes of the kube-apiserver. It can be accessed with an authenticated client, and is enabled by default (Figure 3). These routes for the kube-apiserver are defined within kubernetes/pkg/routes/logs.go (Figure 1), and are mounted in k8s.io/kubernetes/pkg/master/master.go (Figure 2). However, this configuration allows an attacker without host privileges to access protected host logs.

The kubelet has this same functionality, allowing an authenticated user to access the log routes. These routes for the kubelet are defined within kubernetes/pkg/kubelet/kubelet.go (Figure 4). An example of accessing the /var/log directory can be seen in Figure 5.

Parent directory traversal was attempted against these routes on the kubelet and kube-apiserver, but attempts were unsuccessful due to a mitigation in http.ServeFile (Figure 6). The mitigation function, containsDotDot (Figure 7), checks the request path for any presence of “..”. If any are present, the file will not be returned. 

```
type Logs struct{}

func (l Logs) Install(c *restful.Container) {
	// use restful: ws.Route(ws.GET("/logs/{logpath:*}").To(fileHandler))
	// See github.com/emicklei/go-restful/blob/master/examples/restful-serve-static.go
	ws := new(restful.WebService)
	ws.Path("/logs")
	ws.Doc("get log files")
	ws.Route(ws.GET("/{logpath:*}").To(logFileHandler).Param(ws.PathParameter("logpath", "path to the log").DataType("string")))
	ws.Route(ws.GET("/").To(logFileListHandler))

	c.Add(ws)
}

func logFileHandler(req *restful.Request, resp *restful.Response) {
	logdir := "/var/log"
	actual := path.Join(logdir, req.PathParameter("logpath"))
	http.ServeFile(resp.ResponseWriter, req.Request, actual)
}

func logFileListHandler(req *restful.Request, resp *restful.Response) {
	logdir := "/var/log"
	http.ServeFile(resp.ResponseWriter, req.Request, logdir)
}
```
Figure 17.1: The /logs/ endpoint definition for the kube-apiserver.

```
if c.ExtraConfig.EnableLogsSupport {
  routes.Logs{}.Install(s.Handler.GoRestfulContainer)
}
```
Figure 17.2: The configuration flag which determines whether the /logs/ endpoint is registered to the kube-apiserver’s HTTP endpoints.

```
$ hyperkube kube-apiserver --help | grep logs
      --enable-logs-handler                       If true, install a /logs handler for the apiserver logs. (default true)
```
Figure 17.3: The default configuration for the /logs/ endpoint in the kube-apiserver.

```
func (kl *kubelet) Run(updates <-chan kubetypes.PodUpdate) {
	if kl.logServer == nil {
		kl.logServer = http.StripPrefix("/logs/", http.FileServer(http.Dir("/var/log/")))
	}

```
Figure 17.4: The logServer to use when serving logs from /var/log/ on the kubelet.

```
root@node1:/home/ubuntu# curl -k -H "Authorization: Bearer $MY_TOKEN" "https://172.31.28.169:10250/logs/"
<pre>
<a href="alternatives.log">alternatives.log</a>
<a href="amazon/">amazon/</a>
<a href="apt/">apt/</a>
<a href="auth.log">auth.log</a>
<a href="auth.log.1">auth.log.1</a>
<a href="btmp">btmp</a>
...
```
Figure 17.5: An example listing of /var/log on the kubelet.

```
func ServeFile(w ResponseWriter, r *Request, name string) {
	if containsDotDot(r.URL.Path) {
		// Too many programs use r.URL.Path to construct the argument to
		// serveFile. Reject the request under the assumption that happened
		// here and ".." may not be wanted.
		// Note that name might not contain "..", for example if code (still
		// incorrectly) used filepath.Join(myDir, r.URL.Path).
		Error(w, "invalid URL path", StatusBadRequest)
		return
	}
	dir, file := filepath.Split(name)
	serveFile(w, r, Dir(dir), file, false)
}

```
Figure 17.6: The ServeFile function in the net/http package, using the containsDotDot mitigation to prevent parent directory traversal.

```
func containsDotDot(v string) bool {
	if !strings.Contains(v, "..") {
		return false
	}
	for _, ent := range strings.FieldsFunc(v, isSlashRune) {
		if ent == ".." {
			return true
		}
	}
	return false
}
```
Figure 17.7: The containsDotDot function, which checks for the presence of “..” within request path fields.

**Exploit Scenario**
Alice configures a Kubernetes cluster and attempts to harden the underlying host. Eve gains privileged access to Alice’s Kubernetes cluster, and views the logs across the cluster through the kube-apiserver and kubelet /logs/ HTTP endpoint, gaining access to privileged information the host produces in the /var/log/ directory.

**Recommendation**
Short term, disable the serving of the /var/logs directory by default on the kube-apiserver and kubelet. Restrict the serving of logs to files specified within the kube-apiserver or kubelet configuration. Do not serve entire directories.

Long term, remove the serving of log directories and files. Emphasize the use of host log aggregation and centralization.

**Anything else we need to know?**:

See #81146 for current status of all issues created from these findings.

The vendor gave this issue an ID of TOB-K8S-026 and it was finding 17 of the report.

The vendor considers this issue Medium Severity.

To view the original finding, begin on page 52 of the [Kubernetes Security Review Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Environment**:

- Kubernetes version: 1.13.4

**Comment by warmchang** [comment]
FYI: https://github.com/kubernetes/kubernetes/issues/77182 :smiley:

**Comment by mikedanese** [comment]
I've been waiting for this to pop up in an exploit chain. I'd love to default all debugging handlers to off. Some mitigations short of that:
* do more than containsDotDot and don't follow symlinks.
* do a better job of protecting critical secrets e.g. bind or seal in a tpm, encrypt on disk and store encryption keys in an xattr

cc @awly

**Comment by awly** [comment]
While sealing/encrypting critical secrets is a good idea, I don't think it matters here.
For example, if credential is a bearer token, it's much more likely to be logged post-decryption (e.g. logging request headers, unredacted).

**Comment by cji** [comment]
/remove-lifecycle stale

Any further discussion on this? Should we consider implementing the mitigations @mikedanese suggests?

**Comment by joelsmith** [comment]
/close
Fixed by #87273
